### PR TITLE
Move ansible to python-modules and use toPythonApplication + ansible related packages

### PR DESCRIPTION
**Pull request by costrouc**
###### Motivation for this change

Resolves https://github.com/NixOS/nixpkgs/pull/49764 by moving ansible command to python modules and setting default ansible to python3. This is necessary since several packages depend on ansible being a package. I would appreciate if this could be merged into the upcoming release. @MMesch this adds a new kernel ansible that may be useful for jupyterWith.

I would really appreciate any help with testing this PR as I do not use ansible regularly.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc: @dasJ and others already tagged in PR

### CODE REVIEW

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -912,6 +912,18 @@ in {
 
   allpairspy = callPackage ../development/python-modules/allpairspy { };
 
+  inherit (callPackage ../development/python-modules/ansible { })
+    ansible
+    ansible_2_5
+    ansible_2_6
+    ansible_2_7;
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: What's the argument for having different versions available? We usually prefer not do that for Python packages.

Comment (costrouc) [reply pkgs/top-level/python-packages.nix]: No reason. I would prefer a single version too. The issue is that `all-packages.nix` needs to have multiple versions of ansible. What would you suggest?

File: pkgs/development/python-modules/ansible-lint/default.nix [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  version = "4.1.0";
+
+  src = fetchFromGitHub {
+    owner = "ansible";
+    repo = "ansible-lint";
+    rev = "v${version}";
+    sha256 = "06vys67lgdmd5fjflicd4zrngdzqr6qj90wy33va9ndlkdn5xzcf";
+  };
+
+  buildInputs = [ setuptools_scm ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/ansible-lint/default.nix]:
```suggestion
  nativeBuildInputs = [ setuptools_scm ];
```

File: pkgs/development/python-modules/ansible-lint/default.nix [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+, nose
+, setuptools_scm
+, ruamel_yaml
+}:
+
+buildPythonPackage rec {
+  pname = "ansible-lint";
+  version = "4.1.0";
+
+  src = fetchFromGitHub {
```
Comment (dotlambda) [inline pkgs/development/python-modules/ansible-lint/default.nix]: Why not `fetchPypi`? Please add a comment.

File: pkgs/development/python-modules/ansible-kernel/default.nix [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "a59039a1724c0f4f4435316e2ad3383f2328ae61f190e74414a66cc8c4637636";
+  };
+
+  propagatedBuildInputs = [ ipywidgets six docopt tqdm jupyter psutil pyyaml ansible-runner ansible ];
+
+  postPatch = ''
+   touch LICENSE.md
```
Comment (dotlambda) [inline pkgs/development/python-modules/ansible-kernel/default.nix]: FYI: https://github.com/ansible/ansible-jupyter-kernel/pull/82

File: pkgs/development/python-modules/ansible-kernel/default.nix [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+, tqdm
+, jupyter
+, psutil
+, pyyaml
+, ansible-runner
+, ansible
+, python
+}:
+
+let kernelSpecFile = writeText "kernel.json" (builtins.toJSON {
```
Comment (dotlambda) [inline pkgs/development/python-modules/ansible-kernel/default.nix]:
```suggestion
let
  kernelSpecFile = writeText "kernel.json" (builtins.toJSON {
```

File: pkgs/development/python-modules/ansible-kernel/default.nix [hunk 6]
```diff
@@ -0,0 +1,57 @@
+{ stdenv
```
Comment (dotlambda) [inline pkgs/development/python-modules/ansible-kernel/default.nix]:
```suggestion
{ lib
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by dotlambda** [comment]
@costrouc Thank you very much for taking up this work!

cc @c0bw3b @dasJ

File: pkgs/development/python-modules/ansible-kernel/default.nix (hunk 6 above)
Comment (costrouc) [reply pkgs/development/python-modules/ansible-kernel/default.nix]: fixed.

File: pkgs/development/python-modules/ansible-kernel/default.nix (hunk 5 above)
Comment (costrouc) [reply pkgs/development/python-modules/ansible-kernel/default.nix]: fixed.

File: pkgs/development/python-modules/ansible-kernel/default.nix (hunk 4 above)
Comment (costrouc) [reply pkgs/development/python-modules/ansible-kernel/default.nix]: added note to update when your Pr is mereged. Fixed.

File: pkgs/development/python-modules/ansible-lint/default.nix (hunk 3 above)
Comment (costrouc) [reply pkgs/development/python-modules/ansible-lint/default.nix]: fixed.

File: pkgs/development/python-modules/ansible-lint/default.nix (hunk 2 above)
Comment (costrouc) [reply pkgs/development/python-modules/ansible-lint/default.nix]: fixed.

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (dotlambda) [reply pkgs/top-level/python-packages.nix]: How about using `overridePythonAttrs` in `all-packages.nix`?

Comment (costrouc) [reply pkgs/top-level/python-packages.nix]: I added a fix that I would like to see what you think. `ansible` is only exposed in pythonPackages and overridePythonAttrs is used to expose multiple ansible versions as asked. I did the overrides in a seperate file. pkgs/tools/admin/ansible/default.nix

File: pkgs/tools/admin/ansible/default.nix [hunk 7]
```diff
@@ -1,72 +1,27 @@
...
-, windowsSupport ? false
-}:
+{ python3Packages, fetchurl }:
 
-let
-  generic = { version, sha256, py ? python2 }: py.pkgs.buildPythonPackage rec {
-    pname = "ansible";
-    inherit version;
+rec {
+  ansible = python3Packages.toPythonApplication python3Packages.ansible;
```
Comment (dotlambda) [inline pkgs/tools/admin/ansible/default.nix]:
```suggestion
  ansible = with python3Packages; toPythonApplication ansible;
```

File: pkgs/tools/admin/ansible/default.nix [hunk 8]
```diff
@@ -1,72 +1,27 @@
...
+
+  ansible_2_7 = ansible;
 
-    outputs = [ "out" "man" ];
+  ansible_2_6 = python3Packages.ansible.overridePythonAttrs(old: rec {
```
Comment (dotlambda) [inline pkgs/tools/admin/ansible/default.nix]:
```suggestion
  ansible_2_6 = with python3Packages toPythonApplication (ansible.overridePythonAttrs (old: rec {
```
Not sure if the added parantheses is necessary.

File: pkgs/tools/admin/ansible/default.nix [hunk 9]
```diff
@@ -1,72 +1,27 @@
...
-
-    doCheck = false;
-    dontStrip = true;
-    dontPatchELF = true;
-    dontPatchShebangs = false;
-
-    propagatedBuildInputs = with py.pkgs; [
-      pycrypto paramiko jinja2 pyyaml httplib2 boto six netaddr dnspython jmespath dopy
-    ] ++ stdenv.lib.optional windowsSupport pywinrm;
+  ansible_2_5 = python3Packages.ansible.overridePythonAttrs(old: rec {
```
Comment (dotlambda) [inline pkgs/tools/admin/ansible/default.nix]: same here

File: pkgs/development/python-modules/ansible-kernel/default.nix [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+, psutil
+, pyyaml
+, ansible-runner
+, ansible
+, python
+}:
+
+let
+  kernelSpecFile = writeText "kernel.json" (builtins.toJSON {
+       argv = [ "${python.interpreter}" "-m" "ansible_kernel" "-f" "{connection_file}" ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/ansible-kernel/default.nix]:
```suggestion
    argv = [ "${python.interpreter}" "-m" "ansible_kernel" "-f" "{connection_file}" ];
```

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.ansible python3.pkgs.ansible python2.pkgs.ansible-lint python3.pkgs.ansible-lint python2.pkgs.ansible-kernel python3.pkgs.ansible-kernel

File: pkgs/development/python-modules/ansible-kernel/default.nix (hunk 10 above)
Comment (costrouc) [reply pkgs/development/python-modules/ansible-kernel/default.nix]: Fixed

File: pkgs/tools/admin/ansible/default.nix (hunk 9 above)
Comment (costrouc) [reply pkgs/tools/admin/ansible/default.nix]: Fixed

File: pkgs/tools/admin/ansible/default.nix (hunk 8 above)
Comment (costrouc) [reply pkgs/tools/admin/ansible/default.nix]: Fixed

File: pkgs/tools/admin/ansible/default.nix (hunk 7 above)
Comment (costrouc) [reply pkgs/tools/admin/ansible/default.nix]: Fixed

**Comment by costrouc** [comment]
@GrahamcOfBorg build python2.pkgs.ansible python3.pkgs.ansible python2.pkgs.ansible-lint python3.pkgs.ansible-lint python2.pkgs.ansible-kernel python3.pkgs.ansible-kernel

File: pkgs/top-level/python-packages.nix (hunk 1 above)
Comment (costrouc) [reply pkgs/top-level/python-packages.nix]: Resolved.

**Pull request merged**